craft_archives/repo/apt_key_manager.py

```diff
@@ -22,39 +22,96 @@
 import pathlib
 import subprocess
 import tempfile
-from typing import List, Optional
-
-import gnupg  # type: ignore
+from typing import Iterable, List, Optional, Union
 
 from . import apt_ppa, errors, package_repository
 
 logger = logging.getLogger(__name__)
 
-# Default keyring file to use if none is provided. Temporary until we move to
-# per-repo keyrings.
-_DEFAULT_KEYRING = pathlib.Path("/etc/apt/trusted.gpg.d/snapcraft.gpg")
+# Directory for apt keyrings as recommended by Debian for third-party keyrings.
+_KEYRINGS_PATH = pathlib.Path("/etc/apt/keyrings")
 
 # GnuPG command line options that we always want to use.
 _GPG_PREFIX = ["gpg", "--batch", "--no-default-keyring"]
 
 
 def _gnupg_ring(keyring_file: pathlib.Path) -> str:
-    """return a string specifying that ``keyring_file`` is in the binary OpenGPG format.
+    """Return a string specifying that ``keyring_file`` is in the binary OpenGPG format.
+
     This is for use in ``gpg`` commands for APT-related keys.
     """
     return f"gnupg-ring:{keyring_file}"
 
 
+def _call_gpg(
+    *parameters: str,
+    keyring: Optional[pathlib.Path] = None,
+    base_parameters: Iterable[str] = _GPG_PREFIX,
+    stdin: Optional[bytes] = None,
+) -> bytes:
+    if keyring:
+        command = [*base_parameters, "--keyring", f"gnupg-ring:{keyring}", *parameters]
+    else:
+        command = [*base_parameters, *parameters]
+    logger.debug(f"Executing command: {command}")
+    env = {"LANG": "C.UTF-8"}
+    process = subprocess.run(
+        command,
+        input=stdin,
+        capture_output=True,
+        check=True,
+        env=env,
+    )
+    return process.stdout
+
+
+def _get_key_path(
+    key_id: str,
+    *,
+    is_ascii: bool = False,
+    base_path: pathlib.Path = _KEYRINGS_PATH,
+    prefix: str = "craft-",
+) -> pathlib.Path:
+    """Get a Path object where we would expect to find a key.
+
+    :param key_id: The key ID for the keyring file.
+    :param base_path: The directory for the key.
+    :param prefix: The prefix fer the keyfile
+    :param is_ascii: Whether the file is ASCII-armored (.asc suffix)
+
+    :returns: A Path object matching the expected filename
+    """
+    file_base = prefix + key_id[-8:].upper()
+    return base_path.joinpath(file_base).with_suffix(".asc" if is_ascii else ".gpg")
+
+
+def _gen_gpg_fingerprints(path: Union[str, pathlib.Path]) -> Iterable[str]:
+    """Get the fingerprint of a GPG key by file path.
+
+    :returns: the fingerprint of the first GPG key in the file
+    :raises: FileNotFoundError if the file does not exist
+    """
+    if not pathlib.Path(path).is_file():
+        raise FileNotFoundError(f"GPG key file does not exist: {str(path)}")
+
+    response = _call_gpg("--show-keys", str(path)).splitlines(keepends=False)
+    for index, line in enumerate(
+        response, start=1
+    ):  # index will be the next line's index
+        if line.startswith(b"pub   "):
+            yield response[index].decode().strip()
```

Comment:
Since we already have a list of responses in memory, and the caller just iterates over the generator to obtain a list, wouldn't it be simpler to just return a list?

Reply:
Yep! Honestly, I wrote the generator without noticing that `get_key_fingerprints` existed and then forgot to fully remove it.